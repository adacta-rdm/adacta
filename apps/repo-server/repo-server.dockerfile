# Keep in sync with the version in .github/workflows/ci.yml
FROM node:20.14.0-bullseye-slim

RUN groupadd --gid 2222 adacta \
 && useradd --uid 2222 --gid adacta --shell /bin/bash --create-home adacta

WORKDIR /adacta

RUN chown adacta -R /adacta

COPY --chown=adacta dist/repo-server/* /adacta/

USER adacta

EXPOSE 5000

# Start the server
CMD node main.js
