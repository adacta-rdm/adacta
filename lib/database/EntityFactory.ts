import type { IUserId } from "./Ids";

import { AuthenticatedUserInfo } from "~/apps/repo-server/src/graphql/AuthenticatedUserInfo";
import { entityId } from "~/apps/repo-server/src/utils/entityId";
import type { DrizzleEntity, DrizzleEntityNameId, Metadata } from "~/drizzle/DrizzleSchema";
import { entityByTypeId } from "~/drizzle/DrizzleSchema";
import { Service } from "~/lib/serviceContainer/ServiceContainer";

@Service(AuthenticatedUserInfo)
export class EntityFactory {
	static id<T extends DrizzleEntityNameId>(e: T): DrizzleEntity<T>["id"] {
		const entry = Object.entries(entityByTypeId).find(([, v]) => v === e);
		if (!entry) throw new Error(`No entity with typeid ${e}`);
		return entityId(entry[0]) as DrizzleEntity<T>["id"];
	}

	static create<T extends "User", TProps extends EntityProps<"User">>(
		entityName: T,
		props: TProps
	): TProps & { id: DrizzleEntity<T>["id"] };
	static create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps,
		creatorId: IUserId
	): TProps & Metadata & { id: DrizzleEntity<T>["id"] };
	static create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps,
		creatorId?: IUserId
	) {
		const o = {
			...props,
			id: EntityFactory.id(entityName),
		};

		if (entityName === "User") return o;

		if (!creatorId) {
			throw new Error(`Missing required argument 'creatorId' for entity ${entityName}`);
		}

		return Object.assign(o, EntityFactory.metadata(creatorId));
	}

	constructor({ userId }: AuthenticatedUserInfo) {
		this.userId = userId;
	}

	create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps
	) {
		return EntityFactory.create(entityName, props, this.userId);
	}

	private userId: IUserId;

	private static metadata(metadataCreatorId: IUserId) {
		return {
			metadataCreatorId,
			metadataCreationTimestamp: new Date(),
			metadataDeletedAt: null,
		} satisfies Metadata;
	}
}

type EntityProps<T extends DrizzleEntityNameId> = Omit<
	DrizzleEntity<T>,
	// Remove the id/metadata fields as these are populated by the factory
	"metadataCreatorId" | "metadataCreationTimestamp" | "metadataDeletedAt" | "id"
>;
