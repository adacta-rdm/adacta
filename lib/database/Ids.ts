import type { Opaque } from "type-fest";

export type IDeviceDefinitionId = Opaque<string, "DEVICE_DEFINITION">;
export type IDeviceId = Opaque<string, "DEVICE">;
export type IIdPoolId = Opaque<string, "ID_POOL">;
export type IImportPresetId = Opaque<string, "IMPORT_PRESET">;
export type INameCompositionId = Opaque<string, "NAME_COMPOSITION">;
export type INameCompositionVariableId = Opaque<string, "NAME_COMPOSITION_VARIABLE">;
export type INoteId = Opaque<string, "NOTE">;
export type IProjectId = Opaque<string, "PROJECT">;
export type IPropertyId = Opaque<string, "PROPERTY">;
export type IRepositorySettingId = Opaque<string, "REPOSITORY_SETTING">;
export type IResourceId = Opaque<string, "RESOURCE">;
export type ISampleId = Opaque<string, "SAMPLE">;
export type ISampleRelationId = Opaque<string, "SAMPLE_RELATION">;
export type ITransformationId = Opaque<string, "TRANSFORMATION">;
export type IUserId = Opaque<string, "USER">;

/**
 * Maps the GraphQL __typename to their respective ID types.
 * Note: This is only required for types that are not part of CouchDB or types that are named
 * differently than the "type" field of the CouchDB document.
 */
export interface IIdTypeMap {
	// Register types from auth-server (prisma)
	User: IUserId;
	RemoteRepo: IRepositorySettingId;

	// Register alias types from auth-server
	AuthenticatedUser: IUserId;

	// Register types from repo-server (MikroORM)
	Note: INoteId;

	// Register alias types from repo-server
	ResourceGeneric: IResourceId;
	ResourceImage: IResourceId;
	ResourceImportPreset: IResourceId;
	ResourceTabularData: IResourceId;

	// Register MikroORM types
	NameComposition: INameCompositionId;
	NameCompositionVariable: INameCompositionVariableId;
	NameCompositionVariableStatic: INameCompositionVariableId;
	NameCompositionVariableVariable: INameCompositionVariableId;
}
