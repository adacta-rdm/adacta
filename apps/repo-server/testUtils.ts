import { PGlite } from "@electric-sql/pglite";
import { assertDefined } from "@omegadot/assert";
import { PgDialect } from "drizzle-orm/pg-core";
import { drizzle as pgliteDrizzle } from "drizzle-orm/pglite/driver";
import type { SQL } from "drizzle-orm/sql/sql";

import { AuthenticatedUserInfo } from "~/apps/repo-server/src/graphql/AuthenticatedUserInfo";
import { EntityLoader } from "~/apps/repo-server/src/services/EntityLoader";
import { RepositoryManagerPostgres } from "~/apps/repo-server/src/services/RepositoryManagerPostgres";
import { isEntityId } from "~/apps/repo-server/src/utils/isEntityId";
import type { DrizzleEntity } from "~/drizzle/DrizzleSchema";
import { entityByTypeId } from "~/drizzle/DrizzleSchema";
import { EntityFactory } from "~/lib/database/EntityFactory";
import type { IDeviceId, ISampleId, IUserId } from "~/lib/database/Ids";
import { SilentLogger } from "~/lib/logger/SilentLogger";
import { ServiceContainer } from "~/lib/serviceContainer/ServiceContainer";
import { uuid } from "~/lib/uuid";

/**
 * Wraps a simple integer into a date
 * @param day
 */
export function integerToDate(day: number) {
	return new Date(2020, 0, day);
}

/**
 * Unwraps the date back into an integer (reverse action of `integerToDate`)
 * @param a
 */
export function dateToInteger(a: Date | undefined | null) {
	if (!a) {
		return "now";
	}
	return (a.getTime() - integerToDate(1).getTime()) / 60 / 60 / 24 / 1000 + 1;
}

export class TestDeviceCreator {
	device?: DrizzleEntity<"Device">;
	userId: IUserId;
	properties: DrizzleEntity<"Property">[] = [];

	constructor(user: IUserId) {
		this.userId = user;
	}

	public addDevice(definition: DrizzleEntity<"DeviceDefinition">, deviceName: string) {
		this.device = EntityFactory.create(
			"Device",
			{
				couchId: null,
				shortId: null,
				name: deviceName,
				definitionId: definition.id,
				specifications: [],
				imageResourceIds: [],
				setupDescription: [],
			},
			this.userId
		);
		// Reset collected properties which were for a different device
		this.properties = [];
	}

	public addProperty(
		name: string,
		value: IDeviceId | ISampleId,
		begin: Date,
		end: Date | null = null
	) {
		assertDefined(this.device, "addDevice() needs to be called before calling addProperty()");

		this.properties.push(
			EntityFactory.create(
				"Property",
				{
					name,
					begin,
					end,
					ownerDeviceId: this.device.id,
					deviceId: isEntityId(value, "Device") ? value : null,
					sampleId: isEntityId(value, "Sample") ? value : null,
				},
				this.userId
			)
		);
	}

	public getEntities(): [DrizzleEntity<"Device">, ...DrizzleEntity<"Property">[]] {
		assertDefined(this.device, "TestDeviceCreator: Device not defined");
		return [this.device, ...this.properties];
	}
}

export const createTestDb = async () => {
	const sc = new ServiceContainer();

	const drizzle = pgliteDrizzle(new PGlite());
	const rmp = sc.set(new RepositoryManagerPostgres({} as any));
	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
	rmp.db = () => drizzle as any;
	const el = sc.set(new EntityLoader(drizzle as any));
	await rmp.migrate();
	await rmp.createRepository("test");
	const schema = sc.set(rmp.schema("test"));

	const counter = new Map<string, number>();

	// Override EntityFactory.id to generate deterministic ids
	EntityFactory.id = (e) => {
		const entry = Object.entries(entityByTypeId).find(([, v]) => v === e);
		if (!entry) throw new Error(`No entity with typeid ${e}`);
		const hex = Number(entry[0]).toString(16).padStart(2, "0");

		const count = counter.get(hex) ?? 0;
		counter.set(hex, count + 1);

		const id = `${count}`.padStart(12, "0");
		return `00000000-0000-00${hex}-0000-${id}` as DrizzleEntity<typeof e>["id"];
		// 00000000-0000-x000-y000-000000000000
		//               14   19
	};

	const user = EntityFactory.create("User", {
		mikroORMId: uuid(),
		firstName: "Test",
		lastName: "User",
		email: "test.user@example.com",
		passwordHash: "password",
		salt: "salt",
		locale: "en",
		dateStyle: "short",
		timeStyle: "short",
	});
	await drizzle.insert(schema.User).values(user);

	sc.set(new AuthenticatedUserInfo(user.id, "token"));
	sc.set(new SilentLogger());

	return { drizzle, el, schema, user, sc };
};

// Test UUID use a custom format, where the last part is a simple counter
export function testUUIDtoInteger(id: string) {
	const n = id.lastIndexOf("-");
	return `${Number(id.substring(n + 1))}`;
}

export function debugRawQuery(query: SQL<unknown>) {
	const pgDialect = new PgDialect();
	// eslint-disable-next-line no-console
	console.log(pgDialect.sqlToQuery(query.inlineParams()).sql);
}
