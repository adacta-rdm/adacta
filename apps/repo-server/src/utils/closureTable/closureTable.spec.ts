import { asc, gt } from "drizzle-orm";
import type { PgliteDatabase } from "drizzle-orm/pglite/index";
import { beforeEach, describe, expect, test } from "vitest";

import type { EntityLoader } from "~/apps/repo-server/src/services/EntityLoader";
import {
	closureTableDeleteLeaf,
	closureTableDeleteSubtree,
	closureTableInsert,
	closureTableMoveSubtree,
	closureTableSetParent,
} from "~/apps/repo-server/src/utils/closureTable/closureTableTree";
import { createTestDb, testUUIDtoInteger } from "~/apps/repo-server/testUtils";
import type { DrizzleSchema } from "~/drizzle/DrizzleSchema";
import { EntityFactory } from "~/lib/database/EntityFactory";
import type { IDeviceDefinitionId, IUserId } from "~/lib/database/Ids";

describe("closureTable", () => {
	let drizzle: PgliteDatabase;
	let el: EntityLoader;
	let schema: DrizzleSchema;
	let userId: IUserId;

	const dumpRelations = async () => {
		return (
			(
				await el.drizzle
					.select()
					.from(schema.DeviceDefinitionPaths)
					// Filter out self-links as they only clutter the output
					.where((t) => gt(t.depth, 0))
					// Sort by ancestorId, descendantId to make the output deterministic and ensure
					// that the snapshot is readable
					.orderBy(
						asc(schema.DeviceDefinitionPaths.ancestorId),
						asc(schema.DeviceDefinitionPaths.descendantId)
					)
			).map((r) => {
				return {
					ancestor: testUUIDtoInteger(r.ancestorId),
					descendant: testUUIDtoInteger(r.descendantId),
					depth: r.depth,
				};
			})
		);
	};

	const selfLink = async (id: IDeviceDefinitionId) => {
		return drizzle.insert(schema.DeviceDefinitionPaths).values({
			ancestorId: id,
			descendantId: id,
			depth: 0,
		});
	};

	const createDD = async (name: string) => {
		const e = EntityFactory.create(
			"DeviceDefinition",
			{
				name,
				parentDeviceDefinitionIds: [],
				imageResourceIds: [],
				acceptsUnit: [],
				specifications: [],
			},
			userId
		);
		await drizzle.insert(schema.DeviceDefinition).values(e);
		return e;
	};

	beforeEach(async () => {
		const test = await createTestDb();
		userId = test.user.id;
		drizzle = test.drizzle;
		el = test.el;
		schema = test.schema;
	});

	test("Inserts into closure table", async () => {
		const { DeviceDefinitionPaths } = schema;

		const d0 = await createDD("d0");
		const d1 = await createDD("d1");
		const d2 = await createDD("d2");
		const d3 = await createDD("d3");
		const d4 = await createDD("d4");

		await selfLink(d0.id); // First root must be linked to itself manually (?)

		// 0 -> 1
		await closureTableInsert(el, DeviceDefinitionPaths, d0.id, d1.id);
		expect(await dumpRelations()).toMatchSnapshot();

		// 1 -> 2
		await closureTableInsert(el, DeviceDefinitionPaths, d1.id, d2.id);
		expect(await dumpRelations()).toMatchSnapshot();

		// 2 -> 3
		// 2 -> 4
		await closureTableInsert(el, DeviceDefinitionPaths, d2.id, d3.id);
		await closureTableInsert(el, DeviceDefinitionPaths, d2.id, d4.id);
		expect(await dumpRelations()).toMatchSnapshot();
	});

	test("Move subtree", async () => {
		const { DeviceDefinitionPaths } = schema;

		const d0 = await createDD("d0");
		const d1 = await createDD("d1");
		const d2 = await createDD("d2");
		const d3 = await createDD("d3");
		const d4 = await createDD("d4");
		const d5 = await createDD("d5");

		await selfLink(d0.id); // First root must be linked to itself manually (?)

		// 0 -> 1
		await closureTableInsert(el, DeviceDefinitionPaths, d0.id, d1.id);
		// 1 -> 2
		await closureTableInsert(el, DeviceDefinitionPaths, d1.id, d2.id);
		// 2 -> 3
		await closureTableInsert(el, DeviceDefinitionPaths, d2.id, d3.id);
		// 2 -> 4
		await closureTableInsert(el, DeviceDefinitionPaths, d2.id, d4.id);
		// 4 -> 5
		await closureTableInsert(el, DeviceDefinitionPaths, d4.id, d5.id);

		// Right now 4/5 are children of 2, but we want to move them to be children of 3
		await closureTableMoveSubtree(el, DeviceDefinitionPaths, d4.id, d3.id);

		expect(await dumpRelations()).toMatchSnapshot();
	});

	test("Delete Leaf", async () => {
		const { DeviceDefinitionPaths } = schema;

		const d0 = await createDD("d0");
		const d1 = await createDD("d1");
		const d2 = await createDD("d2");
		const d3 = await createDD("d3");

		await selfLink(d0.id); // First root must be linked to itself manually (?)

		// 0 -> 1
		await closureTableInsert(el, DeviceDefinitionPaths, d0.id, d1.id);
		// 1 -> 2
		await closureTableInsert(el, DeviceDefinitionPaths, d1.id, d2.id);
		// 2 -> 3
		await closureTableInsert(el, DeviceDefinitionPaths, d2.id, d3.id);

		await closureTableDeleteLeaf(el, DeviceDefinitionPaths, d3.id);

		expect(await dumpRelations()).toMatchSnapshot();
	});

	test("Delete subtree", async () => {
		const { DeviceDefinitionPaths } = schema;

		const d0 = await createDD("d0");
		const d1 = await createDD("d1");
		const d2 = await createDD("d2");
		const d3 = await createDD("d3");

		await selfLink(d0.id); // First root must be linked to itself manually (?)

		// 0 -> 1
		await closureTableInsert(el, DeviceDefinitionPaths, d0.id, d1.id);
		// 1 -> 2
		await closureTableInsert(el, DeviceDefinitionPaths, d1.id, d2.id);
		// 2 -> 3
		await closureTableInsert(el, DeviceDefinitionPaths, d2.id, d3.id);

		await closureTableDeleteSubtree(el, DeviceDefinitionPaths, d2.id);

		expect(await dumpRelations()).toMatchSnapshot();
	});

	test("Change parent", async () => {
		const { DeviceDefinitionPaths } = schema;

		const d0 = await createDD("d0");
		const d1 = await createDD("d1");
		const d2 = await createDD("d2");
		const d3 = await createDD("d3");
		const d4 = await createDD("d4");
		const d5 = await createDD("d5");

		const newRoot = await createDD("d6");

		// 0 -> 1
		await closureTableInsert(el, DeviceDefinitionPaths, d0.id, d1.id);
		// 1 -> 2
		await closureTableInsert(el, DeviceDefinitionPaths, d1.id, d2.id);
		// 2 -> 3
		await closureTableInsert(el, DeviceDefinitionPaths, d2.id, d3.id);
		// 3 -> 4
		await closureTableInsert(el, DeviceDefinitionPaths, d3.id, d4.id);
		// 4 -> 5
		await closureTableInsert(el, DeviceDefinitionPaths, d4.id, d5.id);

		// Change parent from existing node to new node
		await closureTableSetParent(d2.id, newRoot.id, el, schema.DeviceDefinitionPaths);
		expect(await dumpRelations()).toMatchSnapshot();

		// Change parent to undefined
		await closureTableSetParent(d3.id, undefined, el, schema.DeviceDefinitionPaths);
		expect(await dumpRelations()).toMatchSnapshot();

		// Change parent from undefined to different root
		const newRoot2 = await createDD("d7");
		await closureTableSetParent(d3.id, newRoot2.id, el, schema.DeviceDefinitionPaths);
		expect(await dumpRelations()).toMatchSnapshot();
	});
});
