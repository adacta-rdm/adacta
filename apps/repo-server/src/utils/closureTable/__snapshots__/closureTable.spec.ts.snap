// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`closureTable > Change parent 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "2",
    "depth": 1,
    "descendant": "3",
  },
  {
    "ancestor": "2",
    "depth": 2,
    "descendant": "4",
  },
  {
    "ancestor": "2",
    "depth": 3,
    "descendant": "5",
  },
  {
    "ancestor": "3",
    "depth": 1,
    "descendant": "4",
  },
  {
    "ancestor": "3",
    "depth": 2,
    "descendant": "5",
  },
  {
    "ancestor": "4",
    "depth": 1,
    "descendant": "5",
  },
  {
    "ancestor": "6",
    "depth": 1,
    "descendant": "2",
  },
  {
    "ancestor": "6",
    "depth": 2,
    "descendant": "3",
  },
  {
    "ancestor": "6",
    "depth": 3,
    "descendant": "4",
  },
  {
    "ancestor": "6",
    "depth": 4,
    "descendant": "5",
  },
]
`;

exports[`closureTable > Change parent 2`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "3",
    "depth": 1,
    "descendant": "4",
  },
  {
    "ancestor": "3",
    "depth": 2,
    "descendant": "5",
  },
  {
    "ancestor": "4",
    "depth": 1,
    "descendant": "5",
  },
  {
    "ancestor": "6",
    "depth": 1,
    "descendant": "2",
  },
]
`;

exports[`closureTable > Change parent 3`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "3",
    "depth": 1,
    "descendant": "4",
  },
  {
    "ancestor": "3",
    "depth": 2,
    "descendant": "5",
  },
  {
    "ancestor": "4",
    "depth": 1,
    "descendant": "5",
  },
  {
    "ancestor": "6",
    "depth": 1,
    "descendant": "2",
  },
  {
    "ancestor": "7",
    "depth": 1,
    "descendant": "3",
  },
  {
    "ancestor": "7",
    "depth": 2,
    "descendant": "4",
  },
  {
    "ancestor": "7",
    "depth": 3,
    "descendant": "5",
  },
]
`;

exports[`closureTable > Delete Leaf 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "0",
    "depth": 2,
    "descendant": "2",
  },
  {
    "ancestor": "1",
    "depth": 1,
    "descendant": "2",
  },
]
`;

exports[`closureTable > Delete subtree 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
]
`;

exports[`closureTable > Inserts into closure table 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
]
`;

exports[`closureTable > Inserts into closure table 2`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "0",
    "depth": 2,
    "descendant": "2",
  },
  {
    "ancestor": "1",
    "depth": 1,
    "descendant": "2",
  },
]
`;

exports[`closureTable > Inserts into closure table 3`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "0",
    "depth": 2,
    "descendant": "2",
  },
  {
    "ancestor": "0",
    "depth": 3,
    "descendant": "3",
  },
  {
    "ancestor": "0",
    "depth": 3,
    "descendant": "4",
  },
  {
    "ancestor": "1",
    "depth": 1,
    "descendant": "2",
  },
  {
    "ancestor": "1",
    "depth": 2,
    "descendant": "3",
  },
  {
    "ancestor": "1",
    "depth": 2,
    "descendant": "4",
  },
  {
    "ancestor": "2",
    "depth": 1,
    "descendant": "3",
  },
  {
    "ancestor": "2",
    "depth": 1,
    "descendant": "4",
  },
]
`;

exports[`closureTable > Move subtree 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "0",
    "depth": 2,
    "descendant": "2",
  },
  {
    "ancestor": "0",
    "depth": 3,
    "descendant": "3",
  },
  {
    "ancestor": "0",
    "depth": 4,
    "descendant": "4",
  },
  {
    "ancestor": "0",
    "depth": 5,
    "descendant": "5",
  },
  {
    "ancestor": "1",
    "depth": 1,
    "descendant": "2",
  },
  {
    "ancestor": "1",
    "depth": 2,
    "descendant": "3",
  },
  {
    "ancestor": "1",
    "depth": 3,
    "descendant": "4",
  },
  {
    "ancestor": "1",
    "depth": 4,
    "descendant": "5",
  },
  {
    "ancestor": "2",
    "depth": 1,
    "descendant": "3",
  },
  {
    "ancestor": "2",
    "depth": 2,
    "descendant": "4",
  },
  {
    "ancestor": "2",
    "depth": 3,
    "descendant": "5",
  },
  {
    "ancestor": "3",
    "depth": 1,
    "descendant": "4",
  },
  {
    "ancestor": "3",
    "depth": 2,
    "descendant": "5",
  },
  {
    "ancestor": "4",
    "depth": 1,
    "descendant": "5",
  },
]
`;
